bpaotu/bpaotu/query.py

```diff
@@ -173,37 +187,64 @@ def _q_all_cached(self, topic, q, mutate_result=None):
             cache.set(key, result)
         return result
 
+
     def matching_sample_ids_and_environment(self):
         q = self._session.query(SampleContext.id, SampleContext.environment_id)
         subq = self._build_taxonomy_subquery()
         q = self._apply_filters(q, subq).order_by(SampleContext.id)
         return self._q_all_cached('matching_sample_ids_and_environment', q)
 
-    def matching_sample_headers(self, required_headers=None):
+
+    def matching_sample_headers(self, required_headers=None, sort_col=None, sort_order=None):
         query_headers = [SampleContext.id, SampleContext.environment_id]
+        joins = [] # Keep track of any foreign ontology classes which may be needed to be joined to.
 
         cache_name = ['matching_sample_headers']
         if required_headers:
             cache_name += required_headers
             for h in required_headers:
-                query_headers.append(getattr(SampleContext, h))
+                if h != '':
+                    col = getattr(SampleContext, h)
+
+                    if hasattr(col, "ontology_class"):
+                        foreign_col = getattr(col.ontology_class, 'value')
+                        query_headers.append(foreign_col)
+                        joins.append(col.ontology_class)
+                    else:
+                        query_headers.append(col)
+
+        if sort_order == 'asc':
+            q = self._session.query(*query_headers).outerjoin(*joins).order_by(query_headers[int(sort_col)])
+
+            cache_name.append(str(query_headers[int(sort_col)]))
+            cache_name.append(sort_order)
+
+        elif sort_order == 'desc':
+            q = self._session.query(*query_headers).outerjoin(*joins).order_by(query_headers[int(sort_col)].desc())
+
+            cache_name.append(str(query_headers[int(sort_col)]))
+            cache_name.append(sort_order)
+        else:
+            q = self._session.query(*query_headers).outerjoin(*joins)
 
```

Comment:
you can probably factor out some of this.

I'd move the common bits to the top:

```
q = self._session.query(*query_headers).outerjoin(*joins)
```

then update q to add the order_by stuff if a sort_order is set.

Reply:
Fixed.